### PR TITLE
[clang] Fix behavior of `__is_trivially_relocatable(volatile int)`

### LINKED ISSUE #77091: `volatile int` is trivially copyable but not trivially relocatable
https://godbolt.org/z/4Y9z7sdEE
This Godbolt shows that clang considers `struct S { volatile int i; };` "trivially relocatable", but `volatile int` is considered "not-trivially relocatable". This is inconsistent. It's also inconsistent with the behavior of `std::is_trivially_copyable_v<volatile int> == true`. `__is_trivially_relocatable` should do the same. Otherwise, we end up with a type that is trivially copyable but not trivially relocatable.

### PR DESCRIPTION
**Pull request by AMP999**
Consistent with `__is_trivially_copyable(volatile int) == true` and `__is_trivially_relocatable(volatile Trivial) == true`, `__is_trivially_relocatable(volatile int)` should also be `true`.

Fixes https://github.com/llvm/llvm-project/issues/77091

[clang] [test] New tests for __is_trivially_relocatable(cv-qualified type)

### CODE REVIEW

File: clang/lib/AST/Type.cpp [hunk 1]
```diff
@@ -2651,6 +2651,8 @@ bool QualType::isTriviallyRelocatableType(const ASTContext &Context) const {
     return false;
   } else if (!BaseElementType->isObjectType()) {
     return false;
+  } else if (BaseElementType.isTriviallyCopyableType(Context)) {
+    return true;
```
Comment (cor3ntin) [inline clang/lib/AST/Type.cpp:2655]: Shouldn't this be after the canPassInRegisters test? at the very least it would be more efficient

Comment (AMP999) [reply clang/lib/AST/Type.cpp]: OK! I changed the order.

**Comment by cor3ntin** [comment]
We probably need a release note for that, don't we?
LGTM otherwise

**Comment by AMP999** [comment]
> We probably need a release note for that, don't we? LGTM otherwise

Sure, I'll add a release note to Clang.

**Review by cor3ntin — APPROVED** [review]
LGTM

**Comment by shafik** [comment]
I think I would like some more eyes on this, I don't know if it is obvious to me what it means to reallocate a volatile object.

**Comment by cor3ntin** [comment]
@erichkeane @AaronBallman

File: clang/lib/AST/Type.cpp [hunk 2]
```diff
@@ -2669,6 +2669,8 @@ bool QualType::isTriviallyRelocatableType(const ASTContext &Context) const {
     return false;
   } else if (const auto *RD = BaseElementType->getAsRecordDecl()) {
     return RD->canPassInRegisters();
+  } else if (BaseElementType.isTriviallyCopyableType(Context)) {
+    return true;
   } else {
     switch (isNonTrivialToPrimitiveDestructiveMove()) {
```
Comment (rjmccall) [inline clang/lib/AST/Type.cpp:2675]: I'm not sure why these are different functions, other than maybe having different preconditions.

Comment (AMP999) [reply clang/lib/AST/Type.cpp:2675]: You mean why is `isNonTrivialToPrimitiveDestructiveMove` different from `isTriviallyRelocatable`? I don't know, but I think it has to do with Objective-C, so I left it alone.

**Comment by cor3ntin** [comment]
@AaronBallman @erichkeane ping :)

**Review by erichkeane — COMMENTED** [review]
I don't have a good idea here... I don't really know what it means to 'relocate' a volatile.

**Comment by AMP999** [comment]
> I don't have a good idea here... I don't really know what it means to 'relocate' a volatile.  

This PR doesn't take a position on what it means to relocate a volatile; it just makes the type-trait return the correct value on cv-qualified types, for consistency with `__is_trivially_copyable` and to implement the current Standard proposals.

**Comment by AMP999** [comment]
@AaronBallman, does this look okay to you? See https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p1144r9.html#abstract-opdef-trivially-relocatable and https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2786r3.pdf#subsection.12.4 which both agree on this point.

**Review by AaronBallman — COMMENTED** [review]
I think the changes make sense in as much as relocatable is supposed to be a subset of copyable (I continue to think `volatile`-qualified objects should not be trivially copyable... because they aren't... but that's WG21's issue, not ours).

I'm adding in the libc++ reviewers because this trait is used by libc++ and there's a potential for this to be an ABI-breaking change. I hope we don't need to add any ABI tags for this, but I wanted to give them a chance to weigh in.

**Comment by philnik777** [comment]
This isn't ABI breaking for us currently (at least in a non-benign way). We only use it to optimize `vector` growing currently, and just define it to `is_trivially_copyable` if `__is_trivially_relocatable` isn't available.

**Comment by AaronBallman** [comment]
> This isn't ABI breaking for us currently (at least in a non-benign way). We only use it to optimize `vector` growing currently, and just define it to `is_trivially_copyable` if `__is_trivially_relocatable` isn't available.

Excellent, thank you for the confirmation!

**Review by AaronBallman — APPROVED** [review]
LGTM

**Comment by AMP999** [comment]
Thanks all for the review! @AaronBallman, Could you please land this patch for me?

**Comment by mordante** [comment]
@AMP999 I wanted to merge this for you but there are merge conflicts. Can you resolve these?

**Comment by AMP999** [comment]
@mordante Thank you! I've resolved merge conflicts. I've also added an NFC commit to remove empty messages from `static_assert`s for consistency with the rest of the file.

File: clang/docs/ReleaseNotes.rst [hunk 3]
```diff
@@ -249,6 +249,12 @@ Bug Fixes in This Version
 
 - Clang now doesn't produce false-positive warning `-Wconstant-logical-operand`
   for logical operators in C23.
+  Fixes (`#64356 <https://github.com/llvm/llvm-project/issues/64356>`_).
+- ``__is_trivially_relocatable`` no longer returns ``true`` for non-object types
+  such as references and functions, and no longer returns ``false`` for volatile-qualified types.
+  Fixes (`#67498 <https://github.com/llvm/llvm-project/issues/67498>`_) and
+  (`#77091 <https://github.com/llvm/llvm-project/issues/77091>`_)
+
```
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:257]: That's should be moved _after_ the next line (and you can use the same `#GHXXXX` syntax )

**Review by cor3ntin — APPROVED** [review]

**Pull request merged**